### PR TITLE
58 Ill-Defined Substance Saving

### LINKED ISSUE #58: Ill-defined compound editing on substance page
**As a** *curator*
**I want** *to be able edit and submit structures on the substance page*
**So that** *the new structures get stored in the database and compound ids*

**Acceptance Criteria:**
- if an ill-defined compound displayed in marvinjs is in any way different than that loaded for the substance, a "Save Structure" button is available
- when save structure button is pressed, if the substance had been loaded with an ill-defined compound, the structure is PATCHED to the api using the CID from which it was originally loaded
-- if the save is successful, the compound id is displayed in the form.
-- if the save is not successful, the user is notified of the error
- when the save structure button is pressed and the substance had been loaded without an ill-defined compound, the structure is POSTED to the api
-- if the save is successful, the compound id is displayed in the form.
-- if the save is not successful, the user is notified of the error

### PR DESCRIPTION
**Pull request by michael-barbour**
closes #58 

This ticket requires a MarvinJS rebuild.  [Information here](https://github.com/Chemical-Curation/chemcurator_vuejs/wiki/Marvin-JS).

Functionality is:

- If an Ill-defined compound is loaded, patch that compound.
- If no compound is loaded, post that compound and reload using the returned CID (to ensure the save worked correctly and pull any defaults into store).

 when the Save Compound button is pressed.  Currently there is no way to save changes to a substance.

This also contains changes to the way modules are saved for compound.  Previously data was separated into attributes and relationships. @debboutr pointed out this strips valuable information for no reason.  As such I've combined the `attributes` and `relationships` keys into the whole response as just `data`.

### CODE REVIEW

File: src/components/MarvinWindow.vue [hunk 1]
```diff
@@ -102,8 +72,10 @@ export default {
     window.addEventListener(
```
Comment (debboutr) [inline src/components/MarvinWindow.vue:72]: I think this event listener is persisting on the window and causing an error in the devtools console
`MarvinWindow.vue?1f14:46 Uncaught TypeError: Cannot read property 'postMessage' of null`
that comes up when trying to call `exportMrvfile`. It has caused my browser to crash, when going to the editor and editing, then navigating away from the substance page. Maybe these need to get deleted in the [destroyed](https://vuejs.org/v2/api/#destroyed) life cycle?

File: src/components/ChemicalEditors.vue [hunk 2]
```diff
@@ -22,6 +22,69 @@ export default {
...
+        this.saveIllDefinedCompound();
+      }
+    },
+    saveIllDefinedCompound() {
+      let compoundId = this.$store.state.compound.illdefinedcompound.data.id;
+      // Generic pot body
+      let requestBody = {
+        type: "illDefinedCompound",
+        attributes: {
+          mrvfile: this.$refs["marvin"].mrvfile
```
Comment (debboutr) [inline src/components/ChemicalEditors.vue:41]: we have a reference to the mrvfile in both `this.$refs.marvin.mrvfile` as well as `this.mrvfile` from w/in the MarvinWindow component. Would it be better to mapState in the MarvinWindow component into the compound store w/ something like `editor_structure` that could be a computed property of either the marvin or ketcher components that we could use to build the `requestBody`?

this may not make sense, but it seems like using a $ref as a computed property is [frowned upon in vuejs](https://vuejs.org/v2/guide/components-edge-cases.html#Accessing-Child-Component-Instances-amp-Child-Elements)

> $refs are only populated after the component has been rendered, and they are not reactive. It is only meant as an escape hatch for direct child manipulation - you should avoid accessing $refs from within templates or computed properties.

[here](https://stackoverflow.com/questions/39035498/vuejs-watch-refs) is an SO post as well, it seems like it is working this way so it may make sense to go with it, but shouldn't we imagine that there will only be one structure string (mol/mrvfile) that is being edited at a time? Wouldn't we want to unbind all of the store attributes to compound when the user flips from marvin to ketcher? It's great to populate all this data into state onload but it begs the question of how to manage it through UX?

I'm not sure of what's best, could the `saveIllDefinedCompound` function just rely on what is in state to dispatch the necessary actions for the post/patch requirements? or do we modify those conditions dependent upon some data/computed props in the component and some from state? I would think it easier to just base it off of state alone, but that may require more API calls.

File: src/components/MarvinWindow.vue (hunk 1 above)
Comment (michael-barbour) [reply src/components/MarvinWindow.vue:72]: I've spent some time looking into this.  I believe there are a few different issues.

1. Your crash was caused by an infinite loop.  The "else" used to call `exportMrvfile()` was meant to ensure the editor never failed to have parity with the data in the editor window.  Essentially it would request the mrvfile and the mrvfile response and force another request.  I can't replicate whatever the issue was that this was attempting to fix.  Thanks for catching this!
2. It looks like there is a lot of junk being thrown out by marvin due to the way we are handling this listener on mount.  Rather than the event listener being attached to the component it's being attached to the window.  This seemed to be handled well by removing the method by destroying it.  I'll need to work more with this as it seems marvin is throwing errors now.

File: src/components/ChemicalEditors.vue (hunk 2 above)
Comment (michael-barbour) [reply src/components/ChemicalEditors.vue:41]: > $refs are only populated after the component has been rendered, and they are not reactive. It is only meant as an escape hatch for direct child manipulation - you should avoid accessing $refs from within templates or computed properties.

Regarding this point,  I'm not using this in or as a watched or computed value (in fact there are no watched or computed values on this component) and I don't need reactivity from the mrvfile's value.  All this `$refs.marvin.mrvfile` does is pull whatever value is currently saved on the marvin component when the save button is clicked.  That value is then used to construct the body of the patch or post request.

If your feeling is that the MarvinWindow and KetcherWindow components should be responsible for saving their own compounds, then I'd be happy to implement it that way (looking at it now, that makes a lot of sense).  

> I'm not sure of what's best, could the saveIllDefinedCompound function just rely on what is in state to dispatch the necessary actions for the post/patch requirements? or do we modify those conditions dependent upon some data/computed props in the component and some from state? I would think it easier to just base it off of state alone, but that may require more API calls.

I'm opposed to having the MarvinWindow component update the store outside POSTing/PATCHing.  The local mrvfile on the MarvinWindow contains all potential changes.  All of the methods on the MarvinWindow component are all tools to help us interact with the iframe.

[Vuex's documentation](https://vuex.vuejs.org/guide/structure.html) weighs in on this too:


```
Vuex doesn't really restrict how you structure your code. Rather, it enforces a set of high-level principles
1. Application-level state is centralized in the store.
2. The only way to mutate the state is by committing mutations, which are synchronous transactions.
3. Asynchronous logic should be encapsulated in, and can be composed with actions.
```

I believe having any component's changes be kept track of using the store would be a violation of their core principles 1 and 2.

If you were to push for having one "state" manage this page.  I feel the compound information would need to be loaded and tracked local to the MarvinWindow and KetcherWindow components.  This would require a rewrite to the search bar on the navbar and require the Vuex separation api from the Vuex actions in the docs linked above.

**TL;DR:**  
We should leave the ref as is or move the ill defined compound save onto the Marvin window, which I am in favor of doing.
Unless we are going to implement large structural changes that might make things more difficult, a local and Vuex state are required for managing the Marvin and Ketcher windows (as well as many others).  I am in favor of changing how we do this, but we should at least do a Monday tech talk on it.

**Comment by michael-barbour** [comment]
I've removed the infinite loop between marvin and our app. 

It looks like if a user:

- loads a substance
- navigates away from the substance page
- then loads the same substance again

everything works but a console.exception is thrown from the editorws window for marvin.  I'm going to disregard this for now as it's not affecting UX and it seems like a minor bug on a rapidly changing page.

As for disabling the "Save Compound" button when a substance is loaded or the editor is blank.  If the editor is blank this is really easy to do, but tracking changes is difficult and outside of the scope of this ticket in my opinion.  The only thing I have to go on is comparing the structures and Marvin adds meta information as attributes of the `<cml>` node that is not guaranteed to stay the same.

I've also renamed "mrvfile" on the MarvinWindow to "localMrvfile" to hopefully impress that this is the component specific instance, not what is in store or in the iframe.

**Comment by debboutr** [comment]
this is looking really good, we'ere only beginning to get into the madness that I imagine this page is going to end up having to deal with. Nice work!
